---
layout: post
title: "NDS41でエラーハンドリングの話を聞いてきた"
date: 2015-04-14
tags: nds
---

NDS41 (http://nagaoka.techtalk.jp/no41) に参加してきました。入学ネタということでVagrantに入学して発表しようとぎりぎりまでやってたんですが、発表できるレベルには到達しなかったので諦めて発表なしで参加しました。

ギリギリすぎて懇親会は締め切られていたのですがご好意で入れてもらえました。@civic さんありがとうございました。

O/Rマッパーの薄い本の話とか全体的に面白かったんで、全体のまとめは @yutacar の togetter がオススメです http://togetter.com/li/807353?page=1

ぼくが今回気になった発表を２つ挙げさせてもらうと、

1. エラーハンドリング・クロニクル: http://nekogata.hatenablog.com/entry/2015/04/11/135231
2. １周遅れのScala入学: http://www.slideshare.net/kkatagiri/nds41

の２つです。

先に 2 について。
2013 年の Scala 入学式（http://nekogata.hatenablog.com/entry/2013/09/28/185217) から時を経ての再入学というスタンスの話でした。Javaで実際にやってきた仕事をScalaで置き換えるにはどうすか、という視点の話（で合ってるのかな？）。

@nemuzuka さんも言っていたけど、違うパラダイムの勉強をするのは、刺激になるので、とても意義のあることだと思います。

## エラーハンドリング

CからScala, Goのエラーハンドリングの話まで出ていてとても面白かったです。
比較してきちんと評価し、それを文章化できるのはすごいです。
こういうところ、猫型さんのこと尊敬する。

Eitherってあんまり知らなかったですが、関数型の人たちはやっぱり面白いなあと思いました。
以下、 http://nekogata.hatenablog.com/entry/2015/04/11/135231 より抜粋。

```
def divide(x:Int, y:Int): Either[String, Int] = {
  if (y == 0) {
    Left("can't divide by zero")
  } else {
    Right(x / y)
  }
}

divide(2, 2) // => Right(1)
divide(0, 0) // => Left("can't divide by zero")

def divideAndDouble(x: Int, y:Int): Either[String, Int] = divide(x, y).right.map(_ * 2)

divideAndDouble(2, 2) // => Right(2)
divideAndDouble(0, 0) // => Left("can't divide by zero")
```

これを Go で素直にやろうと思ったらこうなるかな？
Go はメソッドチェーンでプログラミングするようなことはほとんどないので、返り値が複数であるのは何の問題もない感じがします。逆にメソッドチェーンでプログラミングしたいような言語は Either のような型があった方が便利ですね。

```
package main

import (
  "errors"
  "fmt"
)

func divide(x int, y int) (int, error) {
  if (y == 0) {
    return 0, errors.New("can't divide by zero")
  } else {
    return (x / y), nil
  }
}


func divideAndDouble(x int, y int) (int, error) {
  v,e := divide(x,y)
  if e != nil {
    return 0, e
  }
  return v * 2, nil
}


func main() {
  fmt.Println(divide(2, 2)) // => 1, <nil>
  fmt.Println(divide(0, 0)) // => 0, "can't divide by zero"

  fmt.Println(divideAndDouble(2, 2)) // => 2, <nil>
  fmt.Println(divideAndDouble(0, 0)) // => 0, "can't divide by zero"
}
```

ところでこの２つの返り値を持つようなパターン、Cならポインターを使うと思います。Objective-C は C の制約を多く受ける言語なので同じようにポインターを使ってエラーを取得します。

```
#import <Foundation/Foundation.h>

@interface F: NSObject
+ (int) divideX: (int)x Y: (int) y error: (NSError **)error;
@end

@implementation F
+ (int) divideX:(int)x Y:(int)y error:(NSError **)error {
  if (y == 0) {
    *error = [NSError errorWithDomain: @"error" code: 0 userInfo: nil];
    return 0;
  } else {
    return x/y;
  }
}
@end

int main(void) {
  NSError *err = nil;
  int ret = [F divideX:2 Y:2 error:&err];
  if (err == nil) {
    NSLog(@"divide: %d", ret);
  } else {
    NSLog(@"err: %@", err);
  }

  err = nil;
  ret = [F divideX:0 Y:0 error:&err];
  if (err == nil) {
    NSLog(@"divide: %d", ret);
  } else {
    NSLog(@"err: %@", err);
  }

  return 0;
}
```

ただ、猫型さんの記事にもあるように、Goは返り値を全て受け取ることを強要するし、それをチェックすることを強要します。そういうのは Objective-C にはないので全然安全じゃないです。
そんなわけで、初めて Go のこの文法見た時は「スマートだなあ」と思いました。


Swift なら Generics もあるし同じようにできるだろうと思ったんですが、まだできないみたいです。
https://devforums.apple.com/message/1101833#1101833

いろいろ迷って以下のようになりました。ErrorがStringを持てなくて悲しい感じです。Box化すればできるのですが、そこまでしてやりたくなかったので、いまはこれで精一杯

```
//: Playground - noun: a place where people can play

import UIKit

enum Result<T> {
    case Error
    case Value(T)
    
    func map( blk: (T) -> T ) -> Result<T> {
        switch self {
        case .Error:
            return self
        case .Value(let v):
            return .Value(blk(v))
        }
    }
}

func divide(x:Int, y:Int) -> Result<Int> {
    if (y == 0) {
        return .Error
    } else {
        return .Value(x / y)
    }
}

divide(2, 2) // => Value(1)
divide(0, 0) // => Error

func divideAndDouble(x: Int, y:Int) -> Result<Int>{
    return divide(x, y).map{$0 * 2}
}

let v = divideAndDouble(2, 2) // => Value(2)
let e = divideAndDouble(0, 0) // => Error
```

Swiftもタプル返せるので Go のようなエラーハンドリングも良いと思うんですけど、`if let` と相性悪いので `Either` がまともに使えるようになったら `Either` 使いたいなあ。
